edx_notifications/base_data.py

```diff
@@ -0,0 +1,287 @@
+        should be set it is set
+        """
+
+        self._expected_type = [args[0]] if not isinstance(args[0], list) else args[0]
+        self._default = kwargs.get('default', None)
+        self._data = WeakKeyDictionary()
+
+    def __get__(self, instance, owner):
+        """
+        Descriptor getter
+        """
+        if not instance:
+            # called as a class method, so return the descriptor (ourselves)
+            return self
+
+        return self._data.get(instance, self._default)
+
+    def __set__(self, instance, value):
+        """
+        Descriptor setter. Be sure to enforce type on setting. But None is allowed.
+        """
+
+        value_type = type(value)
+
+        if value and type(value) not in self._expected_type:
+            raise TypeError(
+                (
+                    "Field expected type of '{expected}' got '{got}'"
+                ).format(expected=self._expected_type, got=value_type)
+            )
+        self._data[instance] = value
+
+
+class StringField(TypedField):
+    """
+    Specialized subclass of TypedField(unicode) as a convienence
+    """
+
+    def __init__(self, **kwargs):
+        super(StringField, self).__init__(unicode, kwargs)
+
+    def __set__(self, instance, value):
+        """
+        Check to see if we were passed a str type, and if so
+        coerce it into a unicode
+        """
+
+        if isinstance(value, str):
+            super(StringField, self).__set__(instance, unicode(value))
+        else:
+            super(StringField, self).__set__(instance, value)
+
+
+class IntegerField(TypedField):
+    """
+    Specialized subclass of TypedField(int) as a convienence
+    """
+
+    def __init__(self, **kwargs):
+        super(IntegerField, self).__init__([int, long], kwargs)
+
+
+class DictField(TypedField):
+    """
+    Specialized subclass of TypedField(dict) as a convienence
+    """
+
+    def __init__(self, **kwargs):
+        super(DictField, self).__init__(dict, kwargs)
+
+    @classmethod
+    def to_json(cls, data):
+        """
+        Serialize to json
+        """
+
+        if not data:
+            return None
+
+        _dict = copy.deepcopy(data)
+
+        for key, value in _dict.iteritems():
+            if isinstance(value, datetime):
+                _dict[key] = value.isoformat()
+
+        return json.dumps(_dict)
+
+    @classmethod
+    def from_json(cls, value):
+        """
+        Descerialize from json
+        """
+
+        if not value:
+            return None
+
+        _dict = json.loads(value)
+
+        for key, value in _dict.iteritems():
+            if isinstance(value, basestring):
+                # This could be a datetime posing as a ISO8601 formatted string
+                # we so have to apply some heuristics here
+                # to see if we want to even attempt
+                if value.count('-') == 2 and value.count(':') == 2 and value.count('T') == 1:
```

Comment:
Rather than relying on string introspection, maybe dates should be serialized as a dictionary with a `type` key, or something like that?